File: aicode/minicli4/typescript/src/completion.ts
import fs from "node:fs";
import path from "node:path";

export interface CompletionContext {
  commandNames: string[];
  sessionIds: string[];
  configKeys: string[];
  toolNames: string[];
  projectRoot: string;
}

const SUBS: Record<string, string[]> = {
  "/config": ["get", "set", "list", "reset"],
  "/session": ["list", "new", "switch", "delete", "rename", "current"],
  "/tools": ["enable", "disable"],
  "/agents": ["list", "set"]
};

export function complete(text: string, ctx: CompletionContext): string[] {
  if (!text.startsWith("/")) {
    return [];
  }
  const trailing = text.endsWith(" ");
  const parts = text.trim().length > 0 ? text.trim().split(/\s+/) : [];
  if (parts.length <= 1 && !trailing) {
    const needle = parts[0] ?? "/";
    return ctx.commandNames.filter((name) => name.startsWith(needle)).sort();
  }

  const cmd = parts[0] ?? "";
  const subs = SUBS[cmd];
  if (subs) {
    if (parts.length === 1 && trailing) {
      return [...subs];
    }
    if (parts.length === 2 && !trailing) {
      return subs.filter((s) => s.startsWith(parts[1] ?? "")).sort();
    }
  }

  if (cmd === "/session" && parts[1] === "switch") {
    const needle = trailing ? "" : (parts[2] ?? "");
    return ctx.sessionIds.filter((x) => x.startsWith(needle)).sort();
  }

  if (cmd === "/config" && (parts[1] === "set" || parts[1] === "get")) {
    const needle = trailing ? "" : (parts[2] ?? "");
    return ctx.configKeys.filter((x) => x.startsWith(needle)).sort();
  }

  if (cmd === "/tools" && (parts[1] === "enable" || parts[1] === "disable")) {
    const needle = trailing ? "" : (parts[2] ?? "");
    return ctx.toolNames.filter((x) => x.startsWith(needle)).sort();
  }

  if (["/read", "/write", "/ls", "/grep", "/tree", "/export", "/add"].includes(cmd)) {
    const needle = trailing ? "" : (parts[parts.length - 1] ?? "");
    return completePaths(ctx.projectRoot, needle);
  }

  return [];
}

function completePaths(projectRoot: string, prefix: string): string[] {
  const all: string[] = [];
  const walk = (dir: string): void => {
    const entries = fs.readdirSync(dir, { withFileTypes: true }).slice(0, 120);
    for (const entry of entries) {
      if (entry.name === ".git" || entry.name === "node_modules" || entry.name === ".minicli4") {
        continue;
      }
      const full = path.join(dir, entry.name);
      const rel = path.relative(projectRoot, full).replace(/\\/g, "/");
      all.push(rel);
      if (entry.isDirectory() && rel.split("/").length < 6) {
        walk(full);
      }
      if (all.length > 250) {
        return;
      }
    }
  };
  try {
    walk(projectRoot);
  } catch {
    return [];
  }
  return all.filter((x) => x.startsWith(prefix)).sort();
}

File: aicode/minicli4/cpp/src/completion.cpp
#include "minicli4/completion.hpp"

#include <algorithm>
#include <map>

namespace minicli4 {

namespace {
const std::map<std::string, std::vector<std::string>> kSubs = {
    {"/config", {"get", "set", "list", "reset"}},
    {"/session", {"list", "new", "switch", "delete", "rename", "current"}},
    {"/tools", {"enable", "disable"}},
    {"/agents", {"list", "set"}},
};

std::vector<std::string> complete_paths(const CompletionContext& ctx, const std::string& prefix) {
  std::vector<std::string> all = ctx.project_paths;
  if (all.empty()) {
    for (const auto& entry : std::filesystem::recursive_directory_iterator(ctx.project_root)) {
      const auto rel = std::filesystem::relative(entry.path(), ctx.project_root).generic_string();
      if (rel.rfind(".git/", 0) == 0 || rel.rfind("node_modules/", 0) == 0 || rel.rfind(".minicli4/", 0) == 0) {
        continue;
      }
      all.push_back(rel);
      if (all.size() > 250) {
        break;
      }
    }
  }
  std::vector<std::string> out;
  for (const auto& item : all) {
    if (item.rfind(prefix, 0) == 0) {
      out.push_back(item);
    }
  }
  std::sort(out.begin(), out.end());
  return out;
}

std::vector<std::string> filter_prefix(const std::vector<std::string>& items, const std::string& prefix, bool sorted) {
  std::vector<std::string> out;
  for (const auto& item : items) {
    if (item.rfind(prefix, 0) == 0) {
      out.push_back(item);
    }
  }
  if (sorted) {
    std::sort(out.begin(), out.end());
  }
  return out;
}
}  // namespace

std::vector<std::string> complete(const std::string& text, const CompletionContext& ctx) {
  if (text.empty() || text[0] != '/') {
    return {};
  }
  const bool trailing = !text.empty() && text.back() == ' ';
  std::vector<std::string> parts;
  std::string current;
  for (const char c : text) {
    if (c == ' ') {
      if (!current.empty()) {
        parts.push_back(current);
        current.clear();
      }
      continue;
    }
    current.push_back(c);
  }
  if (!current.empty()) {
    parts.push_back(current);
  }

  if (parts.size() <= 1 && !trailing) {
    const std::string needle = parts.empty() ? "/" : parts.front();
    return filter_prefix(ctx.command_names, needle, true);
  }

  const std::string cmd = parts.empty() ? "" : parts.front();
  const auto sub_it = kSubs.find(cmd);
  if (sub_it != kSubs.end()) {
    if (parts.size() == 1 && trailing) {
      return sub_it->second;
    }
    if (parts.size() == 2 && !trailing) {
      return filter_prefix(sub_it->second, parts[1], true);
    }
  }

  if (cmd == "/session" && parts.size() >= 2 && parts[1] == "switch") {
    const std::string needle = trailing ? "" : (parts.size() > 2 ? parts[2] : "");
    return filter_prefix(ctx.session_ids, needle, true);
  }
  if (cmd == "/config" && parts.size() >= 2 && (parts[1] == "set" || parts[1] == "get")) {
    const std::string needle = trailing ? "" : (parts.size() > 2 ? parts[2] : "");
    return filter_prefix(ctx.config_keys, needle, true);
  }
  if (cmd == "/tools" && parts.size() >= 2 && (parts[1] == "enable" || parts[1] == "disable")) {
    const std::string needle = trailing ? "" : (parts.size() > 2 ? parts[2] : "");
    return filter_prefix(ctx.tool_names, needle, true);
  }

  if (cmd == "/read" || cmd == "/write" || cmd == "/ls" || cmd == "/grep" || cmd == "/tree" || cmd == "/export" || cmd == "/add") {
    const std::string needle = trailing ? "" : (parts.empty() ? "" : parts.back());
    return complete_paths(ctx, needle);
  }

  return {};
}

}  // namespace minicli4
